crypto/evp/p_lib.c

```diff
@@ -1558,12 +1547,32 @@ int EVP_PKEY_up_ref(EVP_PKEY *pkey)
 #ifndef FIPS_MODULE
 void evp_pkey_free_legacy(EVP_PKEY *x)
 {
-    if (x->ameth != NULL) {
-        if (x->ameth->pkey_free != NULL)
-            x->ameth->pkey_free(x);
+    const EVP_PKEY_ASN1_METHOD *ameth = x->ameth;
+    ENGINE *tmpe = NULL;
+
+    if (ameth == NULL && x->legacy_cache_pkey.ptr != NULL)
+        ameth = EVP_PKEY_asn1_find(&tmpe, x->type);
```

Comment:
This seems curious...  why isn't the ameth set when creating the legacy cache pkey?

Reply:
Because we really don't have a "proper" legacy key and we don't want to accidentally end up in the legacy code. We don't need it for anything other than "freeing" so we do it just-in-time.

Reply:
Good point.